## Clustering Titanic data with Meanshift
data = 'https://www.kaggle.com/c/titanic/data' choose train.csv

1. Get the data on people on the Titanic, their class, sex age, ticket price and whether they survived.
2. Load into pandas dataframe
3. Drop the PassengerId, Name, Ticket, Cabin columns from the dataframe
4. Change sex column into 0 or 1
5. One-hot-encode the Embarked column (3 different ports in England)
6. Drop rows with missing values
7. what is the best bandwidth to use for our dataset? Use sklearn
8. Fit data to a meanshift model
9. How many clusters do we get
10. Add a column to the titanic dataframe with the cluster label for each person
11. Get mean values of each cluster group
12. Add a column with the size of each cluster group.
13. Write out conclusion from the aggregated data.

In [14]:
import pandas as pd 
import numpy as np
titanic_data = pd.read_csv('titanic_train.csv')
titanic_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [15]:
# Drop irrelevant columns
titanic_data.drop(['PassengerId','Name','Ticket','Cabin'],'columns',inplace=True)
titanic_data.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

## understanding the data

- **Survived** column tells us whether the Passenger survived the sinking of titanic or not. 0 - did not survive, 1 - survived  
- **Pclass** is which class the passenger was travelling ,i.e 1st ,2nd or 3rd.
- **Sex** male or female
- **Age** How old the passenger is .
- **SibSp and Parch** The number of siblings/parents aboard the titanic.
- **Fare** the price of ticket
- **Embarked** tells where the passenger boarded the ship . (C - Cherbourg, Q - Queenstown,S= Southampton )

In [16]:
from sklearn import preprocessing
# Convert gender to 0 or 1
label_enc =preprocessing.LabelEncoder()
titanic_data['Sex'] = label_enc.fit_transform(titanic_data['Sex'].astype(str))
titanic_data.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    1  22.0      1      0   7.2500        S
1         1       1    0  38.0      1      0  71.2833        C
2         1       3    0  26.0      0      0   7.9250        S
3         1       1    0  35.0      1      0  53.1000        S
4         0       3    1  35.0      0      0   8.0500        S

In [17]:
# One-hot encoding of 'Embarked' with pd.get_dummies
titanic_data = pd.get_dummies(titanic_data,columns=['Embarked'])
titanic_data.head()


Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0         0       3    1  22.0      1      0   7.2500           0           0   
1         1       1    0  38.0      1      0  71.2833           1           0   
2         1       3    0  26.0      0      0   7.9250           0           0   
3         1       1    0  35.0      1      0  53.1000           0           0   
4         0       3    1  35.0      0      0   8.0500           0           0   

   Embarked_S  
0           1  
1           0  
2           1  
3           1  
4           1

In [27]:
# Find missing values in the data and drop those rows:
print('rows before drop n/a',len(titanic_data))
bool_matrix = titanic_data.isnull() # dataframe with True and False values for each cell in the titanic_data
only_null_filter = bool_matrix.any(axis=1) # is there a True value in any column in each row. returns a pandas Series with index matching index of titcanic dataframe
missing = titanic_data[only_null_filter] # show all rows that has one or more null values
titanic_data = titanic_data.dropna()
print('rows after',len(titanic_data))
titanic_data
pd.options.display.max_rows = None # let me see all rows in the dataframe (can be used with columns too)
bool_matrix[:50]

rows before drop n/a 714
rows after 714


Survived  Pclass    Sex    Age  SibSp  Parch   Fare  Embarked_C  \
0      False   False  False  False  False  False  False       False   
1      False   False  False  False  False  False  False       False   
2      False   False  False  False  False  False  False       False   
3      False   False  False  False  False  False  False       False   
4      False   False  False  False  False  False  False       False   
6      False   False  False  False  False  False  False       False   
7      False   False  False  False  False  False  False       False   
8      False   False  False  False  False  False  False       False   
9      False   False  False  False  False  False  False       False   
10     False   False  False  False  False  False  False       False   
11     False   False  False  False  False  False  False       False   
12     False   False  False  False  False  False  False       False   
13     False   False  False  False  False  False  False       False   
14     False   False  False  False  False  False  False       False   
15     False   False  False  False  False  False  False       False   
16     False   False  False  False  False  False  False       False   
18     False   False  False  False  False  False  False       False   
20     False   False  False  False  False  False  False       False   
21     False   False  False  False  False  False  False       False   
22     False   False  False  False  False  False  False       False   
23     False   False  False  False  False  False  False       False   
24     False   False  False  False  False  False  False       False   
25     False   False  False  False  False  False  False       False   
27     False   False  False  False  False  False  False       False   
30     False   False  False  False  False  False  False       False   
33     False   False  False  False  False  False  False       False   
34     False   False  False  False  False  False  False       False   
35     False   False  False  False  False  False  False       False   
37     False   False  False  False  False  False  False       False   
38     False   False  False  False  False  False  False       False   
39     False   False  False  False  False  False  False       False   
40     False   False  False  False  False  False  False       False   
41     False   False  False  False  False  False  False       False   
43     False   False  False  False  False  False  False       False   
44     False   False  False  False  False  False  False       False   
49     False   False  False  False  False  False  False       False   
50     False   False  False  False  False  False  False       False   
51     False   False  False  False  False  False  False       False   
52     False   False  False  False  False  False  False       False   
53     False   False  False  False  False  False  False       False   
54     False   False  False  False  False  False  False       False   
56     False   False  False  False  False  False  False       False   
57     False   False  False  False  False  False  False       False   
58     False   False  False  False  False  False  False       False   
59     False   False  False  False  False  False  False       False   
60     False   False  False  False  False  False  False       False   
61     False   False  False  False  False  False  False       False   
62     False   False  False  False  False  False  False       False   
63     False   False  False  False  False  False  False       False   
66     False   False  False  False  False  False  False       False   

    Embarked_Q  Embarked_S  cluster_group  
0        False       False          False  
1        False       False          False  
2        False       False          False  
3        False       False          False  
4        False       False          False  
6        False       False          False  
7        False       False          False  
8        False       False    

In [26]:
only_null_filter[:10]

0    False
1    False
2    False
3    False
4    False
5     True
6    False
7    False
8    False
9    False
dtype: bool

In [20]:
# what is the best bandwidth to use for our dataset?
# The smaller values of bandwith result in tall skinny kernels & larger values result in short fat kernels.
from sklearn.cluster import estimate_bandwidth
estimate_bandwidth(titanic_data)

30.44675914497196

In [21]:
from sklearn.cluster import MeanShift
analyzer = MeanShift(bandwidth=30) 
analyzer.fit(titanic_data)

MeanShift(bandwidth=30, bin_seeding=False, cluster_all=True, min_bin_freq=1,
          n_jobs=None, seeds=None)

In [22]:
labels = analyzer.labels_
print(labels)
print('\n\n',np.unique(labels))

[0 1 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 0 0 1 1 0 0 0 0 0 0 0 0 0
 0 1 0 1 0 0 0 1 0 1 1 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 1 0 0 0 3 0 0 0 1 0 0
 1 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 3 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0
 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 1 0 0 0 0 1 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 1 4 0 0 1 0 0 0 0 2 2 0 0 0 1
 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 2 3 0 2 2 0 1 1 3 0 0 0 0 0 0 2 2 0 0
 0 0 2 0 0 0 1 0 2 0 1 2 0 0 0 3 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 1
 0 0 2 0 0 3 0 0 3 0 0 1 1 1 0 0 2 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 2 0 0 3 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 1 0 0 0 0 0 1 0 1 2 0 0 0 0 1 2 0 0 1 0
 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 2 1 1 0 0 0 2 0 0 0 0 2 0 0 0 0 1 1
 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 2 0 1 1 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0
 2 0 0 1 0 0 0 0 0 1 0 0 

#### 5 closters in above model

In [23]:
#We will add a new column in dataset which shows the cluster the data of a particular row belongs to.

titanic_data['cluster_group'] = np.nan
data_length=len(titanic_data)
for i in range(data_length): # loop 714 rows
    titanic_data.iloc[i,titanic_data.columns.get_loc('cluster_group')] = labels[i] #set the cluster label on each row

titanic_data.head()


Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0         0       3    1  22.0      1      0   7.2500           0           0   
1         1       1    0  38.0      1      0  71.2833           1           0   
2         1       3    0  26.0      0      0   7.9250           0           0   
3         1       1    0  35.0      1      0  53.1000           0           0   
4         0       3    1  35.0      0      0   8.0500           0           0   

   Embarked_S  cluster_group  
0           1            0.0  
1           0            1.0  
2           1            0.0  
3           1            1.0  
4           1            0.0

In [24]:
titanic_data.describe()

Survived      Pclass         Sex         Age       SibSp       Parch  \
count  714.000000  714.000000  714.000000  714.000000  714.000000  714.000000   
mean     0.406162    2.236695    0.634454   29.699118    0.512605    0.431373   
std      0.491460    0.838250    0.481921   14.526497    0.929783    0.853289   
min      0.000000    1.000000    0.000000    0.420000    0.000000    0.000000   
25%      0.000000    1.000000    0.000000   20.125000    0.000000    0.000000   
50%      0.000000    2.000000    1.000000   28.000000    0.000000    0.000000   
75%      1.000000    3.000000    1.000000   38.000000    1.000000    1.000000   
max      1.000000    3.000000    1.000000   80.000000    5.000000    6.000000   

             Fare  Embarked_C  Embarked_Q  Embarked_S  cluster_group  
count  714.000000  714.000000  714.000000  714.000000     714.000000  
mean    34.694514    0.182073    0.039216    0.775910       0.315126  
std     52.918930    0.386175    0.194244    0.417274       0.690647  
min      0.000000    0.000000    0.000000    0.000000       0.000000  
25%      8.050000    0.000000    0.000000    1.000000       0.000000  
50%     15.741700    0.000000    0.000000    1.000000       0.000000  
75%     33.375000    0.000000    0.000000    1.000000       0.000000  
max    512.329200    1.000000    1.000000    1.000000       4.000000

In [25]:
#Grouping passengers by Cluster
titanic_cluster_data = titanic_data.groupby(['cluster_group']).mean()
#Count of passengers in each cluster
titanic_cluster_data['Counts'] = pd.Series(titanic_data.groupby(['cluster_group']).size())
titanic_cluster_data

Survived    Pclass       Sex        Age     SibSp     Parch  \
cluster_group                                                                
0.0            0.336918  2.525090  0.679211  28.256720  0.439068  0.370968   
1.0            0.611111  1.296296  0.527778  36.148148  0.814815  0.500000   
2.0            0.733333  1.000000  0.366667  32.430667  0.600000  0.866667   
3.0            0.733333  1.000000  0.266667  30.333333  1.000000  1.333333   
4.0            1.000000  1.000000  0.666667  35.333333  0.000000  0.333333   

                     Fare  Embarked_C  Embarked_Q  Embarked_S  Counts  
cluster_group                                                          
0.0             15.434139    0.121864    0.046595    0.831541     558  
1.0             65.622688    0.333333    0.018519    0.629630     108  
2.0            131.183883    0.500000    0.000000    0.500000      30  
3.0            239.991940    0.533333    0.000000    0.466667      15  
4.0            512.329200    1.000000    0.000000    0.000000       3

## Conclusion
- Cluster 0
  - Have 558 passengers
  - Survival rate is 33%(very low) means most of them didn't survive
  - They belong to the lower classes 2nd and 3rd class mostly and are mostly male .
  - The average fare paid is `$15`
- Cluster 1
  - Have 108 passengers
  - Survival rate is 61% means a little more than half of them survived
  - They are mostly from 1st and 2nd class
  - The average fare paid is `$65`  
- Cluster 2 i.e the 3rd Cluster
  - Have 30 passengers
  - Survival rate is 73% means most of them survived
  - They are mostly from 1st class
  - The average fare paid is `$131` (high fare)  
- Cluster 3 i.e the 4th Cluster
  - Have 15 passengers
  - Survival rate is 73% means most of them survived
  - They are mostly from 1st class and are mostly female
  - The average fare paid is  `$239` (which is far higher than the 1st cluster average fare)
- The last cluster has just 3 datapoints so it is not that significant hence we can ignore for data analysis
